tools/kth_threshold.cpp

```diff
@@ -0,0 +1,211 @@
+#include "spdlog/sinks/stdout_color_sinks.h"
+#include "spdlog/spdlog.h"
+
+#include "mappable/mapper.hpp"
+
+#include "app.hpp"
+#include "cursor/max_scored_cursor.hpp"
+#include "index_types.hpp"
+#include "io.hpp"
+#include "query/algorithm.hpp"
+#include "util/util.hpp"
+#include "wand_data_compressed.hpp"
+#include "wand_data_raw.hpp"
+
+#include "query/algorithm.hpp"
+#include "scorer/scorer.hpp"
+
+#include "CLI/CLI.hpp"
+
+using namespace pisa;
+
+template <typename IndexType, typename WandType>
+void kt_thresholds(
+    const std::string& index_filename,
+    const std::string& wand_data_filename,
+    const std::vector<Query>& queries,
+    std::string const& type,
+    ScorerParams const& scorer_params,
+    uint64_t k,
+    bool quantized,
+    std::optional<std::string> pairs_filename,
+    std::optional<std::string> triples_filename)
+{
+    IndexType index;
+    mio::mmap_source m(index_filename.c_str());
+    mapper::map(index, m);
+
+    WandType wdata;
+
+    auto scorer = scorer::from_params(scorer_params, wdata);
+
+    mio::mmap_source md;
+    std::error_code error;
+    md.map(wand_data_filename, error);
+    if (error) {
+        spdlog::error("error mapping file: {}, exiting...", error.message());
+        std::abort();
+    }
+    mapper::map(wdata, md, mapper::map_flags::warmup);
+
+    using Pair = std::set<uint32_t>;
+    std::unordered_set<Pair, boost::hash<Pair>> pairs_set;
+
+    using Triple = std::set<uint32_t>;
+    std::unordered_set<Triple, boost::hash<Triple>> triples_set;
+
+    std::string t;
+    if (pairs_filename) {
+        std::ifstream pin(*pairs_filename);
+        while (std::getline(pin, t)) {
+            std::vector<std::string> p;
+            boost::algorithm::split(p, t, boost::is_any_of(" \t"));
+            pairs_set.insert({(uint32_t)std::stoi(p[0]), (uint32_t)std::stoi(p[1])});
+        }
+        spdlog::info("Number of pairs loaded: {}", pairs_set.size());
+    }
+
+    if (triples_filename) {
+        std::ifstream trin(*triples_filename);
+        while (std::getline(trin, t)) {
+            std::vector<std::string> p;
+            boost::algorithm::split(p, t, boost::is_any_of(" \t"));
+            triples_set.insert(
+                {(uint32_t)std::stoi(p[0]), (uint32_t)std::stoi(p[1]), (uint32_t)std::stoi(p[2])});
+        }
+        spdlog::info("Number of triples loaded: {}", triples_set.size());
+    }
+
+    for (auto const& query: queries) {
+        float threshold = 0;
+
+        auto terms = query.terms;
+        topk_queue topk(k);
+        wand_query wand_q(topk);
+
+        for(auto&& term : terms) {
+            Query q;
+            q.terms.push_back(term);
+            wand_q(make_max_scored_cursors(index, wdata, *scorer, q), index.num_docs());
+            topk.finalize();
+            auto results = topk.topk();
+            topk.clear();
+            float t = 0.0;
+            if (results.size() == k) {
+                t = results.back().first;
+            }
```

Comment:
it could be simplified with `results.size() == k ? results.back().first : 0.0`. We can put that directly into `max` function to avoid awkward variable naming.
```cpp
threshold = std::max(threshold, results.size() == k ? results.back().first : 0.0);
```

Reply:
Actually, now that I think about it, we should be able to access both the threshold and the actual size of the `topk_queue` (if not, we should add that). Then, we avoid copying and make it much cleaner:
```cpp
...
wand_q(make_max_scored_cursors(index, wdata, *scorer, q), index.num_docs());
threshold = std::max(threshold, topk.size() == k ? topk.threshold() : 0.0);
```

Reply:
Same thing for the loops below.